git: worktree_commit, Modify checking empty commit. Fixes #723

onee-only opened the pull request:
Changed condition of checking empty commit from `count of entries == 0` to `createdHash == parentHash`.

There was an edge case where the given commit is initial commit and worktree has no entries. So I made another control branch to check it.

I'm not aware of other edge cases it might have. If there is one, please let me know.

Connected to #723

onee-only commented:
By the way, could you give answer on here? -> https://github.com/go-git/go-git/pull/1045#discussion_r1518686414
I would greatly appreciate it if you could satisfy my curiosity.

pjbgf commented on worktree_commit.go:
```diff
@@ -189,7 +190,21 @@ func (h *buildTreeHelper) BuildTree(idx *index.Index, opts *CommitOptions) (plum
+		parentHash = opts.Parents[0]
+	}
+
+	if treeHash == parentHash && !allowEmpty {
```
We should compare tree hashes against each other (current vs parent). Here we are comparing the tree hash against the parent commit hash.

Please change that, but add this check on the calling method instead.

pjbgf commented on worktree_commit.go:
```diff
@@ -175,7 +173,10 @@ type buildTreeHelper struct {
-	if len(idx.Entries) == 0 && (opts == nil || !opts.AllowEmptyCommits) {
+	allowEmpty := opts != nil && opts.AllowEmptyCommits
+
+	if len(opts.Parents) == 0 && len(idx.Entries) == 0 && !allowEmpty {
+		// Empty initial commit is filtered first
```
What are your thoughts on moving both checks to the calling function instead? 

I am under the impression that empty commit validation would be better suited on that context, than on this.

onee-only replied:
Yeah, I agree with your opinion. I'll change it as you requested.

onee-only commented:
I have some considerations. 
1. Changing the commit msg affects its hash. So, empty commit with only changed msg will be commited without allow-empty flag.
2. Our signing logic creates signature with random source. So, I'm guessing empty commit with signkey will be commited without allow-empty flag.

I think comparing treeHash of parent and current tree is more appropriate than this approach.

What are your thoughts on it?

AriehSchneier commented:
Can you please add a test that tests that an error is returned if you try to add an empty commit to a non-empty repo.

AriehSchneier commented:
> I think comparing treeHash of parent and current tree is more appropriate than this approach.
> 
> What are your thoughts on it?

I think that is what @pjbgf said here: https://github.com/go-git/go-git/pull/1050#discussion_r1521102307
As well as here: https://github.com/go-git/go-git/issues/723#issuecomment-1738222594

And I think the current code will not correctly error for an empty commit on a non-empty repo (hence why I just requested that test)

onee-only commented:
Oh! I should have checked the issue carefully. I made a mistake.
Sorry for taking your time.. 😬

onee-only commented:
I changed it as requested. I expect this is the right approach.

There were test failures as I changed the logic. But I think those were not being tested right. So I changed the tests.

onee-only commented on repository_test.go:
```diff
@@ -103,9 +103,10 @@ func createCommit(c *C, r *Repository) plumbing.Hash {
+		All:               true,
+		Author:            &author,
+		Committer:         &author,
+		AllowEmptyCommits: true,
```
I added AllowEmptyCommits flag because this `createCommit` function is used repeatedly in `TestMergeFF` and `TestMergeFF_Invalid`.

Let me know if we should take different approach.

pjbgf replied:
That's absolutely fine. :thumbsup:

onee-only commented on worktree_commit.go:
```diff
@@ -61,16 +59,40 @@ func (w *Worktree) Commit(msg string, opts *CommitOptions) (plumbing.Hash, error
+	if len(opts.Parents) > 0 {
+		parentCommit, err := w.r.CommitObject(opts.Parents[0])
+		if err != nil {
+			// Parent might not be a commit.
+			if err != plumbing.ErrObjectNotFound {
+				return plumbing.ZeroHash, err
+			}
```
I added these lines because sometimes parent was not a commit hash.

On `worktree_test.go:995` `TestStatusCheckedInBeforeIgnored`,

It threw error: `ErrObjectNotFound` when creating the first commit.

By inspecting opts.Parents[0], I got this hash which I was able to find what it is for: `6ecf0ef2c2dffb796033e5a02219af86ec6584e5`
```go
branch, err := r.Reference("refs/heads/master", false)
c.Assert(err, IsNil)
c.Assert(branch.Hash().String(), Equals, "6ecf0ef2c2dffb796033e5a02219af86ec6584e5")
```

I'm not sure if this is intended. But I added it so I can get any thoughts on it.

pjbgf replied:
The commit [6ecf0ef2c2dffb796033e5a02219af86ec6584e5](https://github.com/git-fixtures/basic/commit/6ecf0ef2c2dffb796033e5a02219af86ec6584e5) exists and should be a valid parent for a new commit when using the "Basic" repo fixture as source.

The test you referred to [is creating a fresh new repo](https://github.com/go-git/go-git/blob/cd6633c3c665ba37f766e96c59c57c58051d0e58/worktree_test.go#L991-L995), so the first commit would not have any parents. Maybe you are referring to another test that failed because of this?

onee-only replied:
I found the problem causing this issue. You and I were both right.

Since `defaultTestCommitOptions` we use in the test is a pointer variable,
https://github.com/go-git/go-git/blob/cd6633c3c665ba37f766e96c59c57c58051d0e58/worktree_test.go#L37

It could be modified in other tests. 
Seems like what made the test fail is this code line down below.
https://github.com/go-git/go-git/blob/cd6633c3c665ba37f766e96c59c57c58051d0e58/options.go#L578

I'll change the `defaultTestCommitOptions` as function returning pointer variable.

onee-only commented on worktree_commit.go:
```diff
@@ -61,16 +59,40 @@ func (w *Worktree) Commit(msg string, opts *CommitOptions) (plumbing.Hash, error
 		return plumbing.ZeroHash, err
 	}
 
+	allowEmpty := opts != nil && opts.AllowEmptyCommits
```
I added this check because of #1051. If you think this is unecessary, I'll remove this line.

pjbgf replied:
Nice one. Please add a "Fixes #1051" at the PR description, so that gets closed automatically when we get this merged.

onee-only replied:
The issue will still exist whether I add this line or not.
I think we shouldn't close that issue with this PR because it happens not only in this function.
This line seems redundant if we are not going to solve this issue in this PR.
I think I should remove it in next push.

pjbgf commented on worktree_test.go:
```diff
@@ -286,16 +289,18 @@ func (s *RepositorySuite) TestPullAdd(c *C) {
 	w, err := r.Worktree()
 	c.Assert(err, IsNil)
-	err = os.WriteFile(filepath.Join(path, "bar"), []byte("bar"), 0755)
+	util.WriteFile(fs, "bar", []byte("bar"), 0755)
```
```suggestion
	err = util.WriteFile(fs, "bar", []byte("bar"), 0755)
```

onee-only replied:
Sorry, didn't see that.

pjbgf commented on worktree_commit_test.go:
```diff
@@ -89,6 +89,54 @@ func (s *WorktreeSuite) TestNothingToCommit(c *C) {
+	w, err := r.Worktree()
+	c.Assert(err, IsNil)
+
+	util.WriteFile(fs, "foo", []byte("foo"), 0644)
```
```suggestion
	err = util.WriteFile(fs, "foo", []byte("foo"), 0644)
	c.Assert(err, IsNil)
```

pjbgf commented on worktree_commit_test.go:
```diff
@@ -164,6 +246,33 @@ func (s *WorktreeSuite) TestCommitAmend(c *C) {
+	err := w.Checkout(&CheckoutOptions{})
+	c.Assert(err, IsNil)
+
+	util.WriteFile(fs, "foo", []byte("foo"), 0644)
```
```suggestion
	err = util.WriteFile(fs, "foo", []byte("foo"), 0644)
	c.Assert(err, IsNil)
```

pjbgf commented on worktree_commit_test.go:
```diff
@@ -113,7 +161,41 @@ func (s *WorktreeSuite) TestCommitParent(c *C) {
+	err := w.Checkout(&CheckoutOptions{})
+	c.Assert(err, IsNil)
+
+	util.WriteFile(fs, "foo", []byte("foo"), 0644)
```
```suggestion
	err = util.WriteFile(fs, "foo", []byte("foo"), 0644)
	c.Assert(err, IsNil)
```

AriehSchneier commented:
What should happen, and is there a test for, a merge commit which leaves the tree in the same state as it was (ie merging 2 branches with the same changes)? (Should that be considered an empty commit?)

onee-only commented:
> What should happen, and is there a test for, a merge commit which leaves the tree in the same state as it was (ie merging 2 branches with the same changes)? (Should that be considered an empty commit?)

I tested the behavior with git cli.

```sh
$ git log --all --decorate --oneline --graph

* 1778bba (main) anything
| * 1044aab (HEAD -> hi) ab
| * aaf0fc5 a
| * 2c9a900 second
| * 34b3b53 first
|/  
* a60d58b Init
* 
```
```sh
$ git diff hi main
-- no output --
```

If you merge `main` into `hi`, git will merge those two without error.
```sh
$ git merge main
Merge made by the 'ort' strategy.
```

Even if the worktree remains the same, seems like we'll need to store metadata of the commit (parent commits, etc).
So I  think we'll have to create a commit if we want to follow the behavior of git cli. 

Since `go-git` only implements fast-foward yet, I think we need to discuss it more when we support more methods.

onee-only commented:
@pjbgf Could you re-run the checks? I'm not sure if this failure is related to this PR.
```
FAIL: <autogenerated>:1: ReceivePackSuite.TestDefaultBranch

/home/runner/work/go-git/go-git/plumbing/transport/test/receive_pack.go:80:
    c.Assert(err, IsNil)
... value *url.Error = &url.Error{Op:"Get", URL:"http://localhost:43743/basic.git/info/refs?service=git-receive-pack", Err:(*net.OpError)(0xc0002[56](https://github.com/go-git/go-git/actions/runs/8654261118/job/23731124667?pr=1050#step:6:57)230)} ("Get \"http://localhost:43743/basic.git/info/refs?service=git-receive-pack\": read tcp [::1]:35226->[::1]:43743: read: connection reset by peer")
```

pjbgf commented:
> What should happen, and is there a test for, a merge commit which leaves the tree in the same state as it was (ie merging 2 branches with the same changes)? (Should that be considered an empty commit?)

@AriehSchneier thanks for helping reviewing this PR. Do you think we need an additional test with a FF merge with a branch that has an empty commit on it? Or are you happy for this to be considered once we implement other merge types?

AriehSchneier commented:
> > What should happen, and is there a test for, a merge commit which leaves the tree in the same state as it was (ie merging 2 branches with the same changes)? (Should that be considered an empty commit?)
> 
> @AriehSchneier thanks for helping reviewing this PR. Do you think we need an additional test with a FF merge with a branch that has an empty commit on it? Or are you happy for this to be considered once we implement other merge types?

Sorry for the slow reply I was on leave and hadn't checked my messages.

I'm happy with whatever you think is best (it's been a little while since I looked at the actual change), as long as when the merge types are added its clear that we need to (or don't need to) touch this code to keep it compliant. That said, if there is a possibility off adding more test cases now, I'm not sure why we wouldn't?

onee-only commented:
I agree that we should add more test cases if there are some. 

However, I'm having some difficulty identifying additional test cases at the moment. Could you please suggest some if you have any?

pjbgf commented:
@onee-only thanks for working on this.

ThinkChaos commented:
Thanks to all that worked on this! Looking forward to the next release :)
